diff --git a/modules/clients/qt5/meson.build b/modules/clients/qt5/meson.build
index 40cc467..29af7ac 100644
--- a/modules/clients/qt5/meson.build
+++ b/modules/clients/qt5/meson.build
@@ -2,6 +2,7 @@ qt5 = import('qt5')
 qt5_deps = dependency('qt5', modules: ['Core', 'Gui', 'Widgets'])
 qt5_core = dependency('Qt5Core')
 qt5_version = qt5_core.version()
+qt5_prefix = qt5_core.get_pkgconfig_variable('prefix')
 qt5_inc = qt5_core.get_pkgconfig_variable('includedir')
 qt5_libdir = qt5_core.get_pkgconfig_variable('libdir')
 
@@ -13,6 +14,9 @@ if run_command('[', '-d', qt5_libdir + '/qt5', ']').returncode() == 0
 # elif fs.is_dir(qt5_libdir + '/qt')
 elif run_command('[', '-d', qt5_libdir + '/qt', ']').returncode() == 0
     qt5_im_module_dir = qt5_libdir + '/qt/plugins/platforminputcontexts'
+# elif fs.is_dir(qt5_libdir + '/plugins')
+elif run_command('[', '-d', qt5_prefix + '/plugins', ']').returncode() == 0
+    qt5_im_module_dir = qt5_prefix + '/plugins/platforminputcontexts'
 else
     error('Can not determine Qt5 plugins/platforminputcontexts directory')
 endif
@@ -26,14 +30,18 @@ moc_files = qt5.preprocess(moc_sources : 'im-nimf-qt5.cpp',
     moc_extra_arguments: ['-DMAKES_MY_MOC_HEADER_COMPILE'],
     include_directories: include_directories(qt5_core_private_inc, qt5_gui_private_inc))
 
+cc = meson.get_compiler('c')
+
 cpp_args = [
     '-DG_LOG_DOMAIN="nimf"',
     '-DNIMF_COMPILATION',
+    '-DUSE_DLFCN',
     '-DQT_NO_KEYWORDS',
     '-fPIC'
 ]
 
 deps = [qt5_deps,
+    cc.find_library('dl'),
     dependency('glib-2.0'),
     dependency('gobject-2.0'),
     dependency('gio-2.0')]
@@ -43,7 +51,6 @@ shared_library('qt5im_nimf',
     include_directories: inc,
     cpp_args : cpp_args,
     cpp_std : 'c++11',
-    link_with: libnimf,
     dependencies : deps,
     install : true,
     install_dir : qt5_im_module_dir)
