diff --git a/src/client/qwaylandintegration.cpp b/src/client/qwaylandintegration.cpp
index 46bef294..845bb038 100644
--- a/src/client/qwaylandintegration.cpp
+++ b/src/client/qwaylandintegration.cpp
@@ -293,12 +293,12 @@ QWaylandDisplay *QWaylandIntegration::display() const
 
 QStringList QWaylandIntegration::themeNames() const
 {
-    return GenericWaylandTheme::themeNames();
+    return QGenericUnixTheme::themeNames();
 }
 
 QPlatformTheme *QWaylandIntegration::createPlatformTheme(const QString &name) const
 {
-    return GenericWaylandTheme::createUnixTheme(name);
+    return QGenericUnixTheme::createUnixTheme(name);
 }
 
 // May be called from non-GUI threads
